core/block/import/pb/converter.go

```diff
@@ -402,29 +430,85 @@ func (p *Pb) updateObjectsIDsInCollection(st *state.State, newToOldIDs map[strin
 	}
 }
 
-// getIDForSubObject preserves original id from snapshot for relations and object types
-func (p *Pb) getIDForSubObject(sn *pb.SnapshotWithType, id string) string {
-	originalId := pbtypes.GetString(sn.Snapshot.Data.Details, bundle.RelationKeyId.String())
-	if strings.HasPrefix(originalId, addr.ObjectTypeKeyToIdPrefix) || strings.HasPrefix(originalId, addr.RelationKeyToIdPrefix) {
-		return originalId
-	}
-	return id
+func (p *Pb) shouldReturnError(req *pb.RpcObjectImportRequest, allErrors *converter.ConvertError, params *pb.RpcObjectImportRequestPbParams) bool {
+	return (!allErrors.IsEmpty() && req.Mode == pb.RpcObjectImportRequest_ALL_OR_NOTHING) ||
+		allErrors.IsNoObjectToImportError(len(params.GetPath()))
 }
 
-// cleanupEmptyBlockMigration is fixing existing pages, imported from Notion
-func (p *Pb) cleanupEmptyBlock(snapshot *pb.SnapshotWithType) {
+func (p *Pb) provideRootCollection(allObjects []*converter.Snapshot, widget *converter.Snapshot, oldToNewID map[string]string) (*converter.Snapshot, error) {
 	var (
-		emptyBlock *model.Block
+		rootObjects         []string
+		objectsTypeToImport widgets.ImportWidgetFlags
+		objectsNotInWidgets []*converter.Snapshot
 	)
+	if widget != nil {
+		objectsTypeToImport, rootObjects = p.getObjectsFromWidgets(widget, oldToNewID)
```

Comment:
If we assume that widget.ImportWidgetFlags is extendible structure, that could be extended by some new built-in widget options, than it is better to rename all variable occurences to _flags_ or _widgetFlags_ and make structure method
`func (f *ImportWidgetFlags) IsEmpty() bool { return !f.ImportCollection && !f.ImportSet}`

Otherwise simple list of strings could be passed among functions to provide list of object types and reduce code complexity